File: c/splay_tree.c
/*
* splay_tree.c
*
* C implementation of splay trees for 15-451 (Algorithm Design and Analysis).
*
* For more information, consult the lecture notes:
*
* https://www.cs.cmu.edu/~15451-f22/lectures/lec04-splay.pdf
*
* Yoseph Mak <ymak@andrew.cmu.edu>
* Last update: 8/29/22
*/

#include <stdio.h>
#include <stdlib.h>

// Struct definitions
struct node {
  struct node* parent;
  struct node* left;
  struct node* right;

  int value;
};

struct tree {
  int size;
  struct node* root;
};

// Tree setting functions
/**
 * Sets P's left child to L and updates L's parent to be P if L isn't NULL.
 * 
 * @requires P cannot be NULL
 */
void set_left(struct node *parent, struct node *left) {
  if (parent == NULL) printf("missing node\n");
  parent->left = left;
  if (left != NULL) left->parent = parent;
}

/**
 * Sets P's left child to R and updates R's parent to be P if R isn't null.
 * 
 * @requires P cannot be NULL
 */
void set_right(struct node *parent, struct node *right) {
  if (parent == NULL) printf("missing node\n");
  parent->right = right;
  if (right != NULL) right->parent = parent;
}

/**
 * Changes P's child to be new instead of old. If P is NULL, this instead
 * assumes old is the root and swaps it out accordingly.
 * 
 * @requires old is the root if P is NULL, or else old is a non-NULL child of P
 */
void swap_child(struct tree *T, struct node *parent, struct node *old, struct node *new) {
  if (old == NULL) printf("Can't swap out empty node\n");
  new->parent = parent;

  if (parent == NULL) {
    if (T->root != old) printf("Can't swap out a non-root node at the root\n");
    T->root = new;
  }
  else if (parent->left == old) parent->left = new;
  else if (parent->right == old) parent->right = new;
  else printf("Failed to find %d as a child of %d\n", old->value, parent->value);
}

// Tree rotations
/**
 * 
 *         z                                        z
 *        /       right rotation about y           /
 *       y      ===========================>      x
 *      / \                                      / \
 *     x   C                                    A   y
 *    / \                                          / \
 *   A   B                                        B   C
 *
 */
void rotate_right(struct tree *T, struct node *y) {
  if (y == NULL) printf("Cannot rotate around NULL\n");
  struct node *x = y->left;
  if (x == NULL) printf("No left node found to replace the root\n");
  
  struct node *z = y->parent;
  struct node *A = x->left;
  struct node *B = x->right;
  struct node *C = y->right;

  swap_child(T, z, y, x);
  set_left(x, A);
  set_right(x, y);
  set_left(y, B);
  set_right(y, C);
}

/**
 * 
 *         z                                        z
 *        /                                        /
 *       y                                        x
 *      / \                                      / \
 *     x   C       left rotation about x        A   y
 *    / \       <===========================       / \
 *   A   B                                        B   C
 *
 */
void rotate_left(struct tree *T, struct node *x) {
  if (x == NULL) printf("Cannot rotate around NULL\n");
  struct node *y = x->right;
  if (y == NULL) printf("No right node found to replace the root\n");
  
  struct node *z = x->parent;
  struct node *A = x->left;
  struct node *B = y->left;
  struct node *C = y->right;

  swap_child(T, z, x, y);
  set_right(y, C);
  set_left(y, x);
  set_right(x, B);
  set_left(x, A);
}

// Splay implementation

/**
 * Performs a single splay step on x.
 * 
 * @requires x cannot be NULL
 */
void splay_step(struct tree *T, struct node *x) {
  if (T == NULL) printf("Tree not found\n");
  if (x == NULL) printf("Can't splay NULL\n");

  struct node *y = x->parent;
  if (y == NULL) return; // root case, do nothing

  struct node *z = y->parent;
  if (z == NULL) { // single-rotation (zig) cases
    if (y->left == x) {
      /*
       *                                  y             x
       *   Zig (y is the tree root):     /     ====>     \
       *                                x                 y
       */
      rotate_right(T, y);
      return;
    }
    else if (y->right == x) {
      /*
       *                                y                 x
       *   Zig (y is the tree root):     \     ====>     /
       *                                  x             y
       */
      rotate_left(T, y);
      return;
    }
  }

  // otherwise, we have one of the complex cases (zig-zig or zig-zag)
  if (z->left != NULL && z->left->right == x) {
    /*
     *                    z              z
     *                   /              /             x
     *   Zig-zag:       y     ====>    x   ====>     / \
     *                   \            /             y   z
     *                    x          y
     */
    rotate_left(T, y);
    rotate_right(T, z);
  }
  else if (z->right != NULL && z->right->left == x) {
    /*
     *                  z            z
     *                   \            \               x
     *   Zig-zag:         y   ====>    x   ====>     / \
     *                   /              \           z   y
     *                  x                y
     */
    rotate_right(T, y);
    rotate_left(T, z);
  }
  else if (z->left != NULL && z->left->left == x) {
    /*
     *                    z                         x
     *                   /            y              \
     *   Zig-zig:       y     ====>  / \   ====>      y
     *                 /            x   z              \
     *                x                                 z
     */
    rotate_right(T, z);
    rotate_right(T, y); // at the start of this stage, y is at the top
  }
  else if (z->right != NULL && z->right->right == x) {
    /*
     *                z                                 x
     *                 \              y                /
     *   Zig-zig:       y     ====>  / \   ====>      y
     *                   \          z   x            /
     *                    x                         z
     */
    rotate_left(T, z);
    rotate_left(T, y);
  }
  else {
    printf("Invalid tree\n");
  }
}

/**
 * Splays node x to the top of tre T.
 * 
 * @requires x cannot be NULL and must be in the tree
 */
void splay_node(struct tree *T, struct node *x) {
  if (T == NULL) printf("No tree found\n");
  if (x == NULL) printf("Can't splay NULL\n");
  while (T->root != x) {
    splay_step(T, x);
  }
}

/**
 * Creates a splay tree of size [n] in a linked list structure
 *       n
 *      /
 *    ...
 *    /
 *   1
 *
 * @requires n is nonnegative
 */
struct tree* initialize_tree(int n) {
  if (n < 0) printf("Invalid size\n");
  if (n == 0) return NULL;
  struct tree *T = calloc(1, sizeof(struct tree));
  T->size = n;
  T->root = calloc(n, sizeof(struct node));
  struct node *current = T->root;
  // chain the nodes together
  for (int i = 1; i < n; i++) {
    current->value = i;
    current->parent = current + 1;
    current->parent->left = current;
    current = current->parent;
  }
  // at the end of the loop, current is the nth node
  current->value = n;
  T->root = current;
  return T;
}

/**
 * Splays the node in tree T with value k to the top.
 * 
 * @requires T is a standard initialized tree, i.e. it has
 * nodes numbered from 1 to n, where n is the size.
 */
void splay(struct tree *T, int k) {
  if (T == NULL) printf("Tree cannot be NULL\n");
  if (k <= 0 || k > T->size) printf("Invalid node request\n");
  splay_node(T, T->root + (k - T->root->value));
  if (T->root->value != k) {
    printf("Splay failed\n");
  }
}

int main() {

  struct tree *T = initialize_tree(10);
  splay(T, 1);
  splay(T, 10);
  splay(T, 4);
  splay(T, 7);

  printf("Splay tests passed\n");

  return 0;
}

File: rust/splay_tree.rs
/*
* splay_tree.rs
*
* Rust implementation of a fixed-size splay tree for 15-451 (Algorithm Design and Analysis).
*
* For more information, consult the lecture notes:
*
* https://www.cs.cmu.edu/~15451-f22/lectures/lec04-splay.pdf
*
* Shiva Peri <shivanap@andrew.cmu.edu>
* Last update: 09/18/22
*/

#[derive(Default, Copy, Clone)]
struct Node {
    parent: Option<usize>,  // None, or Some(Index) into Vec<Nodes>
    left: Option<usize>,    // None, or Some(Index) into Vec<Nodes>
    right: Option<usize>    // None, or Some(Index) into Vec<Nodes>
}

#[derive(Clone)]
struct SplayTree {
    root: usize,            // index of root Node in SplayTree::nodes
    nodes: Vec<Node>        // vector of nodes
}

impl SplayTree {

    /**
     * Creates a splay tree of size n whose inorder traversal is
     *
     *  1, 2, 3, ..., n
     *
     * The nodes are arranged in a linked-list-like structure rooted at n:
     *
     *       n
     *      /
     *    ...
     *    /
     *   1
     *
     * The number of nodes in the tree is fixed and can not be modified
     */
    pub fn new(n: usize) -> Self {
        
        let mut nodes = Vec::with_capacity(n);
        for i in 0..n {
            match i {
                0 => nodes.push(Node{ parent: Some(i + 1), left: None, right: None }),
                _ => nodes.push(Node{ parent: Some(i + 1), left: Some(i - 1), right: None })
            }
        }
        nodes[n-1].parent = None;
        
        Self { root: n-1, nodes }
    }

    // for debugging
    pub fn print(&self) {

        println!("\nroot: {} \nnodes:", self.root);

        for i in 0..(self.nodes.len()) {
            let node = &self.nodes[i];

            print!("value: {}, ", i); 
            match node.parent {
                None => print!("parent: none, "),
                Some(x) => print!("parent: {}, ", x)
            }
            match node.left {
                None => print!("left: none, "),
                Some(x) => print!("left: {}, ", x)
            }
            match node.right {
                None => print!("right: none\n"),
                Some(x) => print!("right: {}\n", x)
            }
        }
    }

    // Set the parent of the node at node_idx to the given node
    fn set_parent(&mut self, node_idx: usize, parent_idx: Option<usize>) {
        assert!(node_idx < self.nodes.len());
        let node = &mut self.nodes[node_idx];
        node.parent = parent_idx;
    }

    // Set the left child of the node at node_idx to the given node
    fn set_left(&mut self, node_idx: usize, left_idx: Option<usize>) {

        assert!(node_idx < self.nodes.len());
        let node = &mut self.nodes[node_idx];
        node.left = left_idx;

        match left_idx {
            Some(i) => self.set_parent(i, Some(node_idx)),
            _ => ()
        }
    }

    // Set the right child of the node at node_idx to the given node
    fn set_right(&mut self, node_idx: usize, right_idx: Option<usize>) {

        assert!(node_idx < self.nodes.len());
        let node = &mut self.nodes[node_idx];
        node.right = right_idx;

        match right_idx {
            Some(i) => self.set_parent(i, Some(node_idx)),
            _ => ()
        }
    }

    // Replace whichever child is currently equal to old with new_child
    fn replace_child(&mut self, node_idx: usize, old_child: Option<usize>, new_child: Option<usize>) {

        assert!(old_child != None);
        assert!(new_child != None);
        assert!(node_idx < self.nodes.len());
        let node = &mut self.nodes[node_idx];

        assert!(node.left == old_child || node.right == old_child);
        if node.left == old_child { node.left = new_child; }
        else { node.right = new_child; }
        self.set_parent(new_child.unwrap(), Some(node_idx));
    }    

    /**
     * 
     *         z                                        z
     *        /       right rotation about y           /
     *       y      ===========================>      x
     *      / \                                      / \
     *     x   C                                    A   y
     *    / \                                          / \
     *   A   B                                        B   C
     *
     */
    fn rotate_right(&mut self, y_idx: Option<usize>) {

        let y = &self.nodes[y_idx.unwrap()];     // Cannot rotate around an empty tree
        let x_idx = y.left;
        let x = &self.nodes[x_idx.unwrap()];     // x will replace y's position, so it cannot be null
        
        let z_idx = y.parent;
        let a_idx = x.left;
        let b_idx = x.right;
        let c_idx = y.right;

        self.set_left(x_idx.unwrap(), a_idx);       // x->set_left(A);
        self.set_right(x_idx.unwrap(), y_idx);      // x->set_right(y);
        self.set_left(y_idx.unwrap(), b_idx);       // y->set_left(B);
        self.set_right(y_idx.unwrap(), c_idx);      // y->set_right(C);

        match z_idx {
            None => self.set_root(x_idx),           // y is the root
            Some(i) => self.replace_child(i, y_idx, x_idx)
        }
    }

    /**
     * 
     *         z                                        z
     *        /                                        /
     *       y                                        x
     *      / \                                      / \
     *     x   C       left rotation about x        A   y
     *    / \       <===========================       / \
     *   A   B                                        B   C
     *
     */
    fn rotate_left(&mut self, x_idx: Option<usize>) {

        let x = &self.nodes[x_idx.unwrap()];    // Cannot rotate around an empty tree
        let y_idx = x.right;
        let y = &self.nodes[y_idx.unwrap()];    // y will replace x's position, so it cannot be null
        
        let z_idx = x.parent;
        let a_idx = x.left;
        let b_idx = y.left;
        let c_idx = y.right;

        self.set_right(y_idx.unwrap(), c_idx);      // y->set_right(C);
        self.set_left(y_idx.unwrap(), x_idx);       // y->set_left(x);
        self.set_right(x_idx.unwrap(), b_idx);      // x->set_right(B);
        self.set_left(x_idx.unwrap(), a_idx);       // x->set_left(A);

        match z_idx {
            None => self.set_root(y_idx),           // x is the root
            Some(i) => self.replace_child(i, x_idx, y_idx)
        }
    }

    fn splay_step(&mut self, x_idx: Option<usize>) {
        assert!(x_idx != None);

        let x = &self.nodes[x_idx.unwrap()];
        let y_idx = x.parent;

        if y_idx == None {()}   // root case, do nothing
        let y = &self.nodes[y_idx.unwrap()];
        let z_idx = y.parent;
        
        match z_idx {
            None => {   // single-rotation (zig) cases
                assert!(y.left == x_idx || y.right == x_idx);

                if y.left == x_idx {
                    /*
                    *                                  y             x
                    *   Zig (y is the tree root):     /     ====>     \
                    *                                x                 y
                    */
                    self.rotate_right(y_idx)
                }
                else if y.right == x_idx {
                    /*
                    *                                y                 x
                    *   Zig (y is the tree root):     \     ====>     /
                    *                                  x             y
                    */
                    self.rotate_left(y_idx)
                }
                else {
                    panic!("No case");
                }
            },
            Some(i) => {    // otherwise, we have one of the complex cases (zig-zig or zig-zag)

                let z = &self.nodes[i];

                let z_left = z.left;
                let z_left_left = match z.left {
                    Some(idx) => self.nodes[idx].left,
                    None => None };
                let z_left_right = match z.left {
                    Some(idx) => self.nodes[idx].right,
                    None => None };

                let z_right = z.right;
                let z_right_left = match z.right {
                    Some(idx) => self.nodes[idx].left,
                    None => None };
                let z_right_right = match z.right {
                    Some(idx) => self.nodes[idx].right,
                    None => None };

                assert!((z_left != None && z_left_right == x_idx) ||
                        (z_right != None && z_right_left == x_idx) ||
                        (z_left != None && z_left_left == x_idx) ||
                        (z_right != None && z_right_right == x_idx));
                
                if z_left != None && z_left_right == x_idx {
                    /*
                        *                    z              z
                        *                   /              /             x
                        *   Zig-zag:       y     ====>    x   ====>     / \
                        *                   \            /             y   z
                        *                    x          y
                        */
                    self.rotate_left(y_idx);
                    self.rotate_right(z_idx)
                }
                else if z_left != None && z_left_left == x_idx {
                    /*
                        *                    z                         x
                        *                   /            y              \
                        *   Zig-zig:       y     ====>  / \   ====>      y
                        *                 /            x   z              \
                        *                x                                 z
                        */
                    self.rotate_right(z_idx);
                    self.rotate_right(y_idx)     // at the start of this stage, y is at the top
                } 
                else if z_right != None && z_right_left == x_idx {
                    /*
                        *                  z            z
                        *                   \            \               x
                        *   Zig-zag:         y   ====>    x   ====>     / \
                        *                   /              \           z   y
                        *                  x                y
                        */
                    self.rotate_right(y_idx);
                    self.rotate_left(z_idx)
                }
                else if z_right != None && z_right_right == x_idx {
                    /*
                        *                z                                 x
                        *                 \              y                /
                        *   Zig-zig:       y     ====>  / \   ====>      y
                        *                   \          z   x            /
                        *                    x                         z
                        */
                    self.rotate_left(z_idx);
                    self.rotate_left(y_idx)
                }
                else {
                    panic!("No case");
                }
            }
        }
    }

    // main interface function
    pub fn splay(&mut self, x_idx: Option<usize>) {
        assert!(x_idx != None);
        while self.root != x_idx.unwrap() {
            self.splay_step(x_idx);
        }
    }

    fn set_root(&mut self, x_idx: Option<usize>) {
        assert!(x_idx != None);

        self.root = x_idx.unwrap();
        let x = &mut self.nodes[x_idx.unwrap()];
        x.parent = None;
    }
}

fn main() {
    let mut tree: SplayTree = SplayTree::new(10);
    tree.print();

    println!("splaying 5 ----------");
    tree.splay(Some(5));
    tree.print();
}
